src/strategy/indicators/sar.ts

```diff
@@ -0,0 +1,101 @@
+/**
+ * @description Parabolic SAR
+ * @author @tsrnc2
+ *
+ * @example Formula taken from https://stockcharts.com
+ * Because the formulas for rising and falling SAR are different,
+ * it is easier to divide the calculation into two parts
+ * Rising SAR
+ *
+ * Prior SAR: The SAR value for the previous period.
+ *
+ * Extreme Point (EP): The highest high of the current uptrend.
+ *
+ * Acceleration Factor (AF): Starting at .02, AF increases by .02(accel) each
+ * time the extreme point makes a new high. AF can reach a maximum of
+ *   .20(maxAccel), no matter how long the uptrend extends.
+ *
+ * Current SAR = Prior SAR + Prior AF(Prior EP - Prior SAR)
+ *
+ *   The Acceleration Factor is multiplied by the difference between the
+ * Extreme Point and the prior period's SAR. This is then added to the
+ * prior period's SAR. Note however that SAR can never be above the
+ * prior two periods' lows. Should SAR be above one of those lows,
+ * use the lowest of the two for SAR.
+ *
+ * Falling SAR
+ *
+ * Prior SAR: The SAR value for the previous period.
+ *
+ * Extreme Point (EP): The lowest low of the current downtrend.
+ *
+ * Acceleration Factor (AF): Starting at .02(startAccel), AF increases by .02(accel) each
+ * time the extreme point makes a new low. AF can reach a maximum of
+ *  .20(maxAccel), no matter how long the downtrend extends.
+ *
+ * Current SAR = Prior SAR - Prior AF(Prior SAR - Prior EP)
+ *
+ *   The Acceleration Factor is multiplied by the difference between the
+ * Prior period's SAR and the Extreme Point. This is then subtracted
+ * from the prior period's SAR. Note however that SAR can never be below
+ * the prior two periods' highs. Should SAR be below one of those highs,
+ * use the highest of the two for SAR.
+ */
+export class SAR {
+    private af: number
+    private ep: number
+    private sar: number
+    private uptrend: boolean
+    private startAccel: number
+    private accel: number
+    private maxAccel: number
+    
+    constructor(periods: Record<string, number>[], startAccel: number, accel: number, maxAccel: number) {
```

Comment:
I have here a conceptional question:
notVitaliy told me one day when i was arguing with him about porting indicators, that he wants that the indicator is a stateless calculation and that the strategy, which uses the indicator, has to care for the state. This indicator is holding its state. 
@notVitaliy is this okay for you, or has the old state be added as input from the calling strategy?

Reply:
Im still learning ts but as far as i understand class is a definition, only instantiated when called. The calling party makes a new variable of type SAR and holds its data inside that variable. But i could be very wrong here.

Reply:
See ema calc was example, how current indicators are working:
https://github.com/magic8bot/magic8bot/blob/master/src/strategy/indicators/ema.ts

And i am not saying that its not working (technically). 
Its much easier to implement an indicator, which can store its own state, but notVitality told me that the strategy should store the state and not the indicator, thats just why i am hinting it here.

Reply:
@tsrnc2 I only a class for indicators for the sake of being able to keep their methods under 1 namespace. I know there are better ways to do this, it was just a decision I didn't feel that strongly about correcting once a more final design pattern emerged.

That being said, @Shawn8901 is right. Indicators/analysis should remain stateless. This can be achieved with static methods. They're sometimes known as class level methods as opposed to instance level methods.

```typescript
class EMA {
  public static classMethod() {}
  
  public instanceMethod() {}
}

// No sugar syntax equivalent, TS will yell at you for this one, but it's valid JS
function EMA() {}
EMA.classMethod()
EMA.prototype.instanceMethod() {}

// usage for both

// EMA doesn't need to have been instantiated
EMA.classMethod()

const ema = new EMA()
ema.instanceMethod()
```

Everything's an object in JS/TS. I was just making these classes so that I can just do `import { EMA } from './ema'` and then have all the appropraite method available, instead of `import { calculate } from './ema'`.

The later would become a nightmare to maintain if you needed more than 1 indicator in a strategy and I didn't much like the idea of naming each calculate function individually. A uniform and consistent API was the main goal. Of course, it'd be possible to do something like this and acheive the same effect:

```typescript
const calculate = () => {}

export const EMA = { calculate }
```

I'm all for refactoring to use the syntax above, it's just not a huge prioritity for me right now.